### PR TITLE
add foreach.RunParallel option

### PR DESCRIPTION
**Pull request by dviry**
I've seen multiple questions about the ForEach parallel execution and in my use case the TData object defines whether it should run in parallel or synchronous (think a chain of emails for approvals), so I've played around Decide and ForEach vs While but I found it much easier to simply extend the existing ForEach Primitive with a .RunParallel boolean (by default set to True to keep it backward compatible). I think it's a nice addition to this otherwise great library and maybe beneficial to others as well.

I've added the Mongo serialization (only one tested since that's what I use - could not get the other DB Provider tests to run), a new Sample and a link in the readme.

I hope this satisfies your PR requirements and I do hope to see this soon released (obviously I am open to any feedback!)

Thanks in advance

### CODE REVIEW

File: test/WorkflowCore.IntegrationTests/Scenarios/ForeachSyncScenario.cs [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+        }
+
+        [Fact]
+        public void Scenario()
+        {
+            var workflowId = StartWorkflow(null);
+            WaitForWorkflowToComplete(workflowId, TimeSpan.FromSeconds(30));
+
+            GetStatus(workflowId).Should().Be(WorkflowStatus.Complete);
+            GetData(workflowId).Counter.Should().Be(3);
```
Comment (danielgerlag) [inline test/WorkflowCore.IntegrationTests/Scenarios/ForeachSyncScenario.cs:60]: Can we assert that it was sync?

File: src/WorkflowCore/Primitives/Foreach.cs [hunk 2]
```diff
@@ -8,28 +8,45 @@ namespace WorkflowCore.Primitives
...
+                    return ExecutionResult.Branch(new List<object>(values), new IteratorPersistenceData() { ChildrenActive = true });
+                }
+                else
+                {
+                    return ExecutionResult.Branch(new List<object>(new object[] { values.ElementAt(0) }), new IteratorPersistenceData() { ChildrenActive = true });
+                }
             }
 
-            if (context.PersistenceData is ControlPersistenceData)
+            if (context.PersistenceData is IteratorPersistenceData persistanceData && persistanceData?.ChildrenActive == true)
```
[the remark this answers is not in the document]
Comment (dviry) [reply src/WorkflowCore/Primitives/Foreach.cs:29]: you are right! I'll check instead for ControlPersistenceData and only for IteratorPersistenceData if .RunParallel is set to False

[the remark this answers is not in the document]
Comment (danielgerlag) [reply src/WorkflowCore/Primitives/Foreach.cs:29]: You could also test what the type is, then we simplify by not maintaining 2 persistence types moving forward.

File: src/WorkflowCore/Services/FluentBuilders/StepBuilder.cs [hunk 3]
```diff
@@ -306,6 +306,25 @@ private WorkflowStep IterateParents(int id, string name)
             return stepBuilder;
         }
 
+        public IContainerStepBuilder<TData, Foreach, Foreach> ForEach(Expression<Func<TData, IEnumerable>> collection, Expression<Func<TData, bool>> runParallel)
```
[the remark this answers is not in the document]
Comment (dviry) [reply src/WorkflowCore/Services/FluentBuilders/StepBuilder.cs:309]: of course that would be a nice global feature but I am not sure how would work in a .While (that does not actually pass any Context.Item) - that would require more refactoring and could be done in a separate PR. 
Also I was thinking of somehow passing the .Index property (like in a for loop instead a foreach) - sometimes might be useful to know which item index is actually being processed (i.e. first, last, somewhere in between...); but that would require a separate .Context.ItemOrder property which I did not want to add to the base class.

File: test/WorkflowCore.IntegrationTests/Scenarios/ForeachSyncScenario.cs (hunk 1 above)
Comment (dviry) [reply test/WorkflowCore.IntegrationTests/Scenarios/ForeachSyncScenario.cs:60]: only way I could find to test it is with the Sleep and Counter - the test will fail is .runParallel is set to True. Do you have any better ideas how to actually Assert that?

File: src/WorkflowCore/Primitives/Foreach.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (dviry) [reply src/WorkflowCore/Primitives/Foreach.cs]: I hope my latest push matches what you had in mind... I did not want to overcomplicate the code so I kept the two separated.

**Pull request merged**